custom-elements/ElementInternals-role.html

```diff
@@ -0,0 +1,150 @@
+            constructor() {
+                super();
+                this.internals_ = this.attachInternals();
+                this.internals_.role = role;
+                this.internals_.ariaLabel = "Test label";
+            }
+        });
+    }
+}
+
+function generate_role_element(role) {
+    conditionally_register_element(role);
+    const el = document.createElement(`${role}-element`);
+    document.body.append(el);
+    return el;
+}
+
+function generate_child_role_element(role, parentElement) {
+    conditionally_register_element(role);
+    const el = document.createElement(`${role}-element`);
+    parentElement.append(el);
+    return el;
+}
+
+const test_roles = [
+    'alert',
+    'alertdialog',
+    'application',
+    'article',
+    'banner',
+    'button',
+    'cell',
+    'checkbox',
+    'combobox',
+    'complementary',
+    'contentinfo',
+    'definition',
+    'dialog',
+    ['directory', 'list'], // is this expected?
+    'document',
+    'figure',
+    'form',
+    'grid',
+    'group',
+    'heading',
+    ['img', 'image'],
+    'feed',
+    'link',
+    'list',
+    'listbox',
+    'log',
+    'main',
+    'marquee',
+    'math',
+    'menu',
+    'menubar',
+    'meter',
+    'navigation',
+    'note',
+    'progressbar',
+    'radio',
+    'radiogroup',
+    'region',
+    'scrollbar',
+    'search',
+    'searchbox',
+    'separator',
+    'slider',
+    'spinbutton',
+    'status',
+    'switch',
+    'table',
+    'tablist',
+    'tabpanel',
+    'term',
+    'textbox',
+    'timer',
+    'toolbar',
+    'tooltip',
+    'tree',
+    'treegrid'
+];
+
+test_roles.map((testdata) => {
+    let role;
+    let roleName;
+    if (Array.isArray(testdata)) {
+        role = testdata[0];
+        roleName = testdata[1];
+    } else {
+        role = testdata;
+        roleName = testdata;
+    }
+    promise_test(async () => {
+        const el = generate_role_element(role);
+        const computedRole = await test_driver.get_computed_role(el);
+        assert_equals(computedRole, roleName, el);
+        el.remove();
+    }, `Applies "${role}" via Element Internals`);
+});
+
+const child_test_roles = [
+    ['table', 'row', 'columnheader'],
+    ['grid', 'row', 'gridcell'],
+    ['list', 'listitem'],
+    ['menu', 'menuitem'],
+    ['menu', 'menuitemcheckbox'],
+    ['menu', 'menuitemradio'],
+    ['listbox', 'option'],
+    ['table', 'row'],
+    ['table', 'rowgroup'],
+    ['table', 'row', 'rowheader'],
+    ['tablist', 'tab'],
+    ['tree', 'treeitem'],
+];
+
+child_test_roles.map((roles) => {
+    promise_test(async () => {
+        const els = roles.reduce((acc, role) => {
+            const previous = acc.at(-1);
+            if (previous) {
+                acc.push(generate_child_role_element(role, previous))
+            } else {
+                acc.push(generate_role_element(role));
+            }
+            return acc;
+        }, []);
```

Comment:
I'm having a little difficulty trying to ascertain what this all generates. I wonder if either this code can be refactored for clarity, or if perhaps we can split some of the functionality out into separate tests? (Maybe the table tests can be a different set, for example?)

Reply:
The goal is to establish a parent/child/grandchild relationship between the elements as required by the specific role in question. So any array `['table', 'row', 'rowheader']` will be rendered as `['parent', 'child', 'ancestor']` or:

```
<table-element>
    <row-element>
        <rowheader-element></rowheader-element>
    </row-element>
</table-element>
```

Some browsers allow these elements without ancestor relations, which is why it may be useful to test in the negative if that's not reasonable, however they were still "roles", which lead me to including them here. If I break them out, does it make sense to have something like:
- `ElementInternals-role-simple.html`
- `ElementInternals-role-parent-child.html`
- `ElementInternals-role-complex.html`
Or is there a better idea for clarity?

Reply:
Breaking them into sub tests isn't especially necessary, but I think the combination of `map`/`reduce` makes it difficult to see what tests are being emitted. Would it be possible to instead use static markup and write out tests for each one?

Reply:
Gotcha. Before I go all the way to static markup, how does this newer version hit you? If you're only looking for static markup for the complex relationships, then I could see making that transition here, but it's a good amount of room for error (as there are currently 12 different complex relationship tests) that I'd love to keep away from if that were possible.

Reply:
@domenic I saw you agreed on the map/reduce note. Is the current version better enough or should I move to adding DOM fixtures? Want to do whatever I can to move this towards a merge and highlight the issue as well as hopefully a fix in WebKit!

Reply:
I think I would personally prefer to see static markup as it will be easier to see the complex relationships when debugging this for tests. The more explicit the better, IMO.

Reply:
Gotcha, updated!